### PR TITLE
Provenance Activity work PLFM-1307

### PR DESCRIPTION
**Pull request by dburdick**

### CODE REVIEW

File: lib/jdomodels/src/main/resources/schema/Revision-ddl.sql [hunk 1]
```diff
@@ -10,6 +11,7 @@ CREATE TABLE `JDOREVISION` (
   PRIMARY KEY (`OWNER_NODE_ID`,`NUMBER`),
   UNIQUE KEY `UNIQUE_REVISION_LABEL` (`OWNER_NODE_ID`,`LABEL`),
   KEY `JDOREVISION_KEY_OWN` (`OWNER_NODE_ID`),
+  CONSTRAINT `REVISION_ACTIVITY_ID_FK` FOREIGN KEY (`ACTIVITY_ID`) REFERENCES `ACTIVITY` (`ID`) ON DELETE SET NULL,
```
Comment (dburdick) [inline lib/jdomodels/src/main/resources/schema/Revision-ddl.sql:14]: test delete activity, keeping node and setting id to null

File: lib/jdomodels/src/main/resources/schema/Activity-ddl.sql [hunk 2]
```diff
@@ -0,0 +1,13 @@
+CREATE TABLE `ACTIVITY` (
+  `ID` bigint(20) NOT NULL,
+  `ETAG` char(36) NOT NULL,  
+  `CREATED_BY` bigint(20) NOT NULL,
+  `CREATED_ON` bigint(20) NOT NULL,
+  `MODIFIED_BY` bigint(20) NOT NULL,
+  `MODIFIED_ON` bigint(20) NOT NULL,
+  `SERIALIZED_OBJECT` mediumblob,
```
Comment (dburdick) [inline lib/jdomodels/src/main/resources/schema/Activity-ddl.sql:8]: change to just "BLOB" for 65k of storage

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/provenance/Activity.json [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+				"MANUAL", 
+				"UNDEFINED" 
+			]
+		},
+		"used": {
+			"type": "array",
+			"description": "The entity references used by this Activity.",
+			"uniqueItems": true,
+			"contentEncoding": "binary",
+			"items": {
```
Comment (dburdick) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/provenance/Activity.json:67]: wrap Reference with a Used object that has enum of type (normal, executed_code, etc)

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/provenance/Activity.json [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+				"$ref": "org.sagebionetworks.repo.model.Reference"
+			}
+		},
+        "executedEntity":{
```
Comment (dburdick) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/provenance/Activity.json:71]: remove this in favor of above

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/provenance/Activity.json [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+			"title": "Created By",
+			"transient": true
+		},
+		"modifiedBy": {
+			"type": "string",
+			"description": "The user that last modified this object.",
+			"title": "Modified By",
+			"transient": true
+		},
+		"activityType": {
```
Comment (dburdick) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/provenance/Activity.json:52]: remove until needed

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/DBOActivityDAOImplAutowiredTest.java [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+		assertEquals(1+initialCount, activityDao.getCount()); 
+		toDelete.add(id);
+		assertNotNull(id);
+		
+		// This activity should exist & make sure we can fetch it		
+		Activity loaded = activityDao.get(id);
+		assertEquals(id, loaded.getId().toString());
+		assertNotNull(loaded.getEtag());
+		
+		checkMigrationDependenciesNoParentDistinctModifier(id);
```
Comment (dburdick) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/DBOActivityDAOImplAutowiredTest.java:131]: break out into own test and test corner cases of pagination

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOActivityDAOImpl.java [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+	@Override
+	public List<Reference> getEntitiesGeneratedBy(String id) {
+		Map<String, Object> parameters = new HashMap<String, Object>();
+		parameters.put(COL_REVISION_ACTIVITY_ID, id);
+		List<Reference> generatedBy = null;
+		generatedBy = simpleJdbcTemplate.query(SQL_GET_NODES_FOR_ACTIVITY, new RowMapper<Reference>() {
+			@Override
+			public Reference mapRow(ResultSet rs, int rowNum)
+					throws SQLException {
+				BigDecimal targetId = rs.getBigDecimal(COL_REVISION_OWNER_NODE);
```
Comment (dburdick) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOActivityDAOImpl.java:255]: - change to getLong(...) and check/set null
- also paginate 
- also return just a list of node ids (Long) and join to currentRev in Node table

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOActivityDAOImpl.java [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+	@Override
+	public void delete(String id) throws DatastoreException, NotFoundException {
+		if(!doesActivityExist(id)) throw new NotFoundException("Activity with id " + id + " could not be found.");
+		MapSqlParameterSource param = new MapSqlParameterSource();
+		param.addValue(COL_ACTIVITY_ID.toLowerCase(), id);
+		basicDao.deleteObjectById(DBOActivity.class, param);
+	}
+	
+	@Override
+	public List<String> getIds() {
```
Comment (dburdick) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOActivityDAOImpl.java:145]: delete this

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOActivityDAOImpl.java [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+	@Autowired
+	private TagMessenger tagMessenger;
+	@Autowired
+	private DBOBasicDao basicDao;	
+	@Autowired
+	private SimpleJdbcTemplate simpleJdbcTemplate;	
+	
+	private static final String SELECT_ALL_IDS_SQL = "SELECT " + SqlConstants.COL_ACTIVITY_ID + " FROM " + SqlConstants.TABLE_ACTIVITY;
+
+	private static final String SELECT_FOR_RANGE_SQL = "SELECT * FROM " + TABLE_ACTIVITY
```
Comment (dburdick) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOActivityDAOImpl.java:60]: select id and etag only

**Comment by dburdick** [comment]
add test to lockActivityAndIncrementEtag in ActivityManagerImplTest to make sure etag is updated.

refactor to not update twice

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOActivityDAOImpl.java [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+			SimpleJdbcTemplate simpleJdbcTemplate) {
+		super();
+		this.tagMessenger = tagMessenger;
+		this.basicDao = basicDao;
+		this.simpleJdbcTemplate = simpleJdbcTemplate;
+	}
+
+	@Transactional(readOnly = false, propagation = Propagation.REQUIRED)
+	@Override
+	public <T extends Activity> T create(T dto) throws DatastoreException, InvalidModelException {
```
Comment (dburdick) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOActivityDAOImpl.java:86]: just return the new id

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOActivityDAOImpl.java [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+			throws InvalidModelException, NotFoundException,
+			ConflictingUpdateException, DatastoreException {
+		return update(dto, true);
+	}
+
+	/**
+	 * @param fromBackup Whether we are updating from backup.
+	 *                   Skip optimistic locking and accept the backup e-tag when restoring from backup.
+	 */
+	private <T extends Activity> T update(T dto, boolean fromBackup) throws DatastoreException,
```
Comment (dburdick) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOActivityDAOImpl.java:114]: when fromBackup is true send eTag message

add lock without etag change method (that sends message)

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOActivityDAOImpl.java [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+		if (!success) throw new DatastoreException("Unsuccessful updating user Activity in database.");
+		T updatedActivity = ActivityUtils.copyDboToDto(dbo);
+
+		return updatedActivity;
+	} // the 'commit' is implicit in returning from a method annotated 'Transactional'
+	
+	
+	@Override
+	public Activity get(String id) throws DatastoreException, NotFoundException {
+		if(!doesActivityExist(id)) throw new NotFoundException("Activity with id " + id + " could not be found.");
```
Comment (dburdick) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOActivityDAOImpl.java:130]: remove this and move message to convert getDBO's NotFoundException to this message

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/ActivityManagerImpl.java [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+		populateCreationFields(userInfo, activity);
+		// set default type
+		if(activity.getActivityType() == null) activity.setActivityType(ActivityType.UNDEFINED);
+		return activityDAO.create(activity);
+	}
+	
+	@Transactional(readOnly = false, propagation = Propagation.REQUIRED)
+	@Override
+	public <T extends Activity> T updateActivity(UserInfo userInfo, T activity)
+			throws InvalidModelException, NotFoundException,
```
Comment (dburdick) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/ActivityManagerImpl.java:66]: enforce that user can not change createdBy or ModifiedBy

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOActivityDAOImpl.java [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+		DBOActivity dbo = getDBO(id);		
+		return ActivityUtils.copyDboToDto(dbo);
+	}
+	
+	@Transactional(readOnly = false, propagation = Propagation.REQUIRED)
+	@Override
+	public void delete(String id) throws DatastoreException, NotFoundException {
```
Comment (dburdick) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOActivityDAOImpl.java:137]: send delete message

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/ActivityManagerImpl.java [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+	public boolean doesActivityExist(String id) {
+		return activityDAO.doesActivityExist(id);
+	}
+	
+	
+	/*
+	 * Private Methods
+	 */
+	private static void populateCreationFields(UserInfo userInfo, Activity a) {
+		Date now = new Date();
```
Comment (dburdick) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/ActivityManagerImpl.java:131]: make default privacy and test with unit test (for both methods)

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/NodeDAOImpl.java [hunk 16]
```diff
@@ -1030,6 +1035,22 @@ public Long getCurrentRevisionNumber(String nodeId) throws NotFoundException, Da
 	}
 	
 	@Override
+	public String getActivityId(String nodeId) throws NotFoundException, DatastoreException {
+		return getActivityId(nodeId, getCurrentRevisionNumber(nodeId));	
+	}
+
+	@Override
+	public String getActivityId(String nodeId, Long revNumber) throws NotFoundException, DatastoreException {
+		if(nodeId == null) throw new IllegalArgumentException("Node Id cannot be null");
```
Comment (dburdick) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/NodeDAOImpl.java:1044]: make sure revision is not null

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/NodeManagerImpl.java [hunk 17]
```diff
@@ -124,6 +124,15 @@ public String createNewNode(Node newNode, UserInfo userInfo)  throws DatastoreEx
 			throw new UnauthorizedException(userInfo.getUser().getUserId()+" is not allowed to create items of type "+newNode.getNodeType());
 		}
 
+		// check whether the user is allowed to connect to the specified activity
```
Comment (dburdick) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/NodeManagerImpl.java:127]: refactor out into method

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/NodeManagerImpl.java [hunk 18]
```diff
@@ -520,4 +539,33 @@ public boolean doesNodeHaveChildren(String nodeId) {
...
+		return activityManager.getActivity(userInfo, activityId);
+	}
+
+	@Transactional(readOnly = false, propagation = Propagation.REQUIRED)
+	@Override
+	public void setActivityForNode(UserInfo userInfo, String nodeId,
+			String activityId) throws NotFoundException, UnauthorizedException,
+			DatastoreException {
+		Node toUpdate = get(userInfo, nodeId);
+		toUpdate.setActivityId(activityId);
```
Comment (dburdick) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/NodeManagerImpl.java:558]: enforce logic: does user have permission to view activity

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/TempMockAuthDao.java [hunk 19]
```diff
@@ -42,6 +42,12 @@ public boolean canCreate(UserInfo userInfo, Node node) {
 		return true;
 	}
 
+	@Override
```
Comment (dburdick) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/TempMockAuthDao.java:45]: this whole class should probably be deleted

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/NodeManagerImpleUnitTest.java [hunk 20]
```diff
@@ -182,6 +186,126 @@ public void testCreateNode() throws Exception {
...
+	@Test
+	public void testCreateNodeActivity404() throws Exception {
+		// Test creating a new node with nothing but the name and type set
+		Node newNode = new Node();
+		newNode.setName("testCreateNode");
+		newNode.setNodeType(EntityType.folder.name());
+		when(mockAuthDao.canCreate(eq(mockUserInfo), (Node)any())).thenReturn(true);
+		String activityId = "8439208403928402";
+		newNode.setActivityId(activityId);
+		when(mockEntityBootstrapper.getChildAclSchemeForPath("/root")).thenReturn(ACL_SCHEME.INHERIT_FROM_PARENT);
```
Comment (dburdick) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/NodeManagerImpleUnitTest.java:198]: remove line

File: services/repository/src/main/java/org/sagebionetworks/repo/web/controller/EntityController.java [hunk 21]
```diff
@@ -848,4 +852,101 @@ PresignedUrl getAttachmentUrl(
...
+	@RequestMapping(value = { 
+			UrlHelpers.ENTITY_GENERATED_BY
+			}, method = RequestMethod.GET)
+	public @ResponseBody
+	Activity getActivityForEntity(
+			@PathVariable String id,
+			@RequestParam(value = AuthorizationConstants.USER_ID_PARAM, required = false) String userId,
+			HttpServletRequest request)
+			throws NotFoundException, DatastoreException, UnauthorizedException {
+		return serviceProvider.getEntityService().getActivityForEntity(userId, id, null, request);
```
Comment (dburdick) [inline services/repository/src/main/java/org/sagebionetworks/repo/web/controller/EntityController.java:876]: make a method signature without version number

File: services/repository/src/test/java/org/sagebionetworks/repo/web/controller/EntityControllerTest.java [hunk 22]
```diff
@@ -240,21 +240,21 @@ public void testPLFM_1288() throws Exception{
...
 		Code two = new Code();
 		two.setName("code");
 		two.setParentId(one.getId());
 		two.setEntityType(Code.class.getName());
 		try{
-			two = (Code) entityServletHelper.createEntity(two, TEST_USER1);
+			two = (Code) entityServletHelper.createEntity(two, TEST_USER1, null);
 			fail("Code cannot have a parent of type Study");
 		}catch(IllegalArgumentException e){
 			System.out.println(e.getMessage());
```
Comment (dburdick) [inline services/repository/src/test/java/org/sagebionetworks/repo/web/controller/EntityControllerTest.java:260]: add test for non null activity id

File: services/repository-managers/src/main/resources/managers-spb.xml [hunk 23]
```diff
@@ -197,6 +207,7 @@
 				<ref bean="userGroupDAO"/>
 				<ref bean="nodeDao"/>
 				<ref bean="accessRequirementDAO"/>
```
Comment (dburdick) [inline services/repository-managers/src/main/resources/managers-spb.xml:209]: move both access requirement and activity above node
